src/include/scoring.h

```diff
@@ -27,224 +27,81 @@
-
-#include <algorithm>
-#include <cmath>
-#include <future>
-#include <iostream>
-#include <limits>
-#include <memory>
-#include <numeric>
-#include <queue>
-#include <ranges>
-#include <set>
-#include <span>
-// #include <execution>
-
-#include "detail/linalg/linalg_defs.h"
-#include "tdb_defs.h"
-#include "utils/fixed_min_heap.h"
-#include "utils/timer.h"
+// @todo Implement
+// #include "detail/scoring/cosine.h"
+// #include "detail/scoring/cosine_avx.h"
 
-#include "utils/print_types.h"
+#ifdef TILEDB_VS_ENABLE_BLAS
 
-// ----------------------------------------------------------------------------
-// Helper utilities
-//----------------------------------------------------------------------------
+// @todo Implement
+#include "detail/scoring/cosine_blas.h"
+#include "detail/scoring/inner_product_blas.h"
+#include "detail/scoring/l2_distance_blas.h"
+
+#endif  // TILEDB_VS_ENABLE_BLAS
+
+#include "detail/linalg/matrix.h"
+#include "utils/fixed_min_heap.h"
+#include "utils/utils.h"
+
+// Helper utility
 namespace {
 class with_ids {};
 class without_ids {};
 }  // namespace
 
-// ----------------------------------------------------------------------------
-// Distance functions
-// ----------------------------------------------------------------------------
+/****************************************************************
+ *
+ * Function objects.  We put these here instead of in the detail
+ * header file so that we can specify which functio we want to
+ * dispatch to in this header
+ *
+ ****************************************************************/
 
-/**
- * @brief Compute sum of squares distance between two vectors.
- * @tparam V
- * @tparam U
- * @param a
- * @param b
- * @return
- */
-#if 0
-template <class V, class U>
-inline auto sum_of_squares(V const& a, U const& b) {
-  float sum{0.0};
-  size_t size_a = size(a);
-
-  if constexpr (std::is_same_v<decltype(a[0]),decltype(b[0])>) {
-    for (size_t i = 0; i < size_a; ++i) {
-      float diff = a[i]- b[i];
-      sum += diff * diff;
-    }
-  } else {
-    for (size_t i = 0; i < size_a; ++i) {
-      float diff = ((float)a[i]) - ((float)b[i]);
-      sum += diff * diff;
-    }
-  }
-  return sum;
-}
-#else
-template <class V, class U>
-inline auto sum_of_squares(V const& a, U const& b) {
-  float sum{0.0};
-  size_t size_a = size(a);
-
-  if constexpr (
-      std::unsigned_integral<std::remove_reference_t<decltype(a[0])>> ||
-      std::unsigned_integral<std::remove_reference_t<decltype(b[0])>>) {
-    for (size_t i = 0; i < size_a; ++i) {
-      // float diff = (float)a[i] - (float)b[i];  // converting to float is slow
-      float diff = (float)a[i] - (float)b[i];
-      sum += diff * diff;
-    }
-  } else {
-    for (size_t i = 0; i < size_a; ++i) {
-      // float diff = (float)a[i] - (float)b[i];  // converting to float is slow
-      float diff = a[i] - b[i];
-      sum += diff * diff;
-    }
-  }
-  return sum;
-}
+namespace _l2_distance {
```

Comment:
Is the plan that we will also have a `_inner_product_distance` namespace here with similar function? Is there a reason we can't do it today? I'm trying to understand what we need beyond this PR to get us to a point where we can expose an "which distance function to use" option to the Python API, perhaps you could help explain that a bit more please?

Reply:
Yes, this will be in the next PR (add `Distance` parameter to all functions computing a distance). I will also sketch out how to interface to Python with the new distance parameterization.